contracts/smart-account/modules/AccountRecoveryModule.sol

```diff
@@ -0,0 +1,528 @@
+import {ECDSA} from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import {IAccountRecoveryModule} from "../interfaces/IAccountRecoveryModule.sol";
+
+/**
+ * @title Account Recovery module for Biconomy Smart Accounts.
+ * @dev Compatible with Biconomy Modular Interface v 0.1
+ *         - It allows the submission and execution of recovery requests
+ *         - EOA guardians only
+ *         - For security reasons the guardian's address is not stored,
+ *           instead, its signature over CONTROL_HASH is used
+ *         - Security delay is always applied to new guardians and recovery requests
+ *         - It is highly recommended to not set security delay to 0
+ * @dev For the validation stage (validateUserOp) can not use custom errors
+ * as EntryPoint contract doesn't support custom error messages within its
+ * 'try IAccount(sender).validateUserOp catch Error' expression
+ * so it becomes more difficult to debug validateUserOp if it uses custom errors
+ * For the execution methods custom errors are used
+ *
+ * @author Fil Makarov - <filipp.makarov@biconomy.io>
+ * based on https://vitalik.ca/general/2021/01/11/recovery.html by Vitalik Buterin
+ */
+
+contract AccountRecoveryModule is
+    BaseAuthorizationModule,
+    IAccountRecoveryModule
+{
+    using ECDSA for bytes32;
+
+    string public constant NAME = "Account Recovery Module";
+    string public constant VERSION = "0.1.0";
+
+    // keccak256(abi.encodePacked("ACCOUNT RECOVERY GUARDIAN SECURE MESSAGE"))
+    bytes32 public constant CONTROL_HASH =
+        0x1c3074d7ce4e35b6f0a67e3db0f31ff117ca8f0a1853e86a95837fea35af828b;
+
+    // guardianID => (smartAccount => TimeFrame)
+    // guardianID = keccak256(signature over CONTROL_HASH)
+    // complies with associated storage rules
+    // see https://eips.ethereum.org/EIPS/eip-4337#storage-associated-with-an-address
+    // see https://docs.soliditylang.org/en/v0.8.15/internals/layout_in_storage.html#mappings-and-dynamic-arrays
+    mapping(bytes32 => mapping(address => TimeFrame)) internal _guardians;
+
+    mapping(address => SaSettings) internal _smartAccountSettings;
+
+    mapping(address => RecoveryRequest) internal _smartAccountRequests;
+
+    /**
+     * @dev Initializes the module for a Smart Account.
+     * Can only be used at a time of first enabling the module for a Smart Account.
+     * @param guardians the list of guardians
+     * @param timeFrames validity timeframes for guardians
+     * @param recoveryThreshold how many guardians' signatures are required to authorize recovery request
+     * @param securityDelay amount of time required to pass between the submission of the recovery request
+     * and its execution
+     * @dev no need for explicit check `length == 0` as it is covered by `recoveryThreshold > length` and
+     * `recoveryThreshold == 0` cheks. So length can never be 0 while recoveryThreshold is not 0
+     */
+    function initForSmartAccount(
+        bytes32[] memory guardians,
```

Comment:
is it cheaper to `memory` arrays here rather than `calldata`?

Reply:
`TimeFrame[]` can't be `calldata` here, but for `guardians` you're right, it will be probably cheaper to make it `calldata`

Reply:
> TimeFrame[] can't be calldata here
Why is this the case?